### PR TITLE
Backward compatibility for mmap payload storage

### PR DESCRIPTION
**Pull request by coszio**
Adds the integration necessary to be able to load and use a segment with mmap storage, but never actually creates it. 
The intention is so that we can release a patch version which can handle a rollback from a newer minor version, e.g. to let `1.13.0` able to be rolled back into `1.12.3`, if the new storage turns out to be buggy.
 
Also centralizes the `on_disk_payload` translation into `PayloadStorageType` so that it becomes easier to update in the future.

Just for reference, [this](https://github.com/qdrant/qdrant/blob/d5b7f22748b2784956fd6df4ba70394c05b3649c/lib/segment/src/segment_constructor/segment_constructor_base.rs#L690-L692) is where the segment config is read. This PR makes it able to handle the `PayloadStorageType::Mmap` variant.

**Note**: Thanks to #5190 (@agourlay) from which I basically git-checkout'ed the integration part

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -4,19 +4,19 @@ version = 3
 
 [[package]]
 name = "actix-codec"
-version = "0.5.0"
```
Comment (agourlay) [inline Cargo.lock:7]: there are a lot of unrelated dependency changes here, did you run cargo update on the all repo?

File: lib/segment/src/payload_storage/mmap_payload_storage.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    fn wipe(&mut self) -> OperationResult<()> {
+        self.storage.write().wipe();
+        Ok(())
+    }
+
+    fn flusher(&self) -> Flusher {
+        let storage = self.storage.clone();
+        Box::new(move || {
+            storage.write().flush().unwrap(); // TODO: error conversion
```
Comment (agourlay) [inline lib/segment/src/payload_storage/mmap_payload_storage.rs:144]: I think we need to address this TODO

File: lib/segment/src/payload_storage/mmap_payload_storage.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    fn iter<F>(&self, mut callback: F) -> OperationResult<()>
+    where
+        F: FnMut(PointOffsetType, &Payload) -> OperationResult<bool>,
+    {
+        self.storage.read().iter(|point_id, payload| {
+            match callback(point_id, payload) {
+                Ok(true) => Ok(true),
+                Ok(false) => Ok(false),
+                Err(e) => {
+                    // TODO return proper error
```
Comment (agourlay) [inline lib/segment/src/payload_storage/mmap_payload_storage.rs:158]: I think we need to address this TODO

**Comment by agourlay** [comment]
Thanks for pushing the integration forward, I added a few comments that I try to handle myself.

File: lib/common/common/src/cpu.rs [hunk 4]
```diff
@@ -208,7 +208,7 @@ pub enum ThreadPriorityError {
     #[error("Failed to set thread priority: {0:?}")]
     SetThreadPriority(thread_priority::Error),
     #[error("Failed to parse thread priority value: {0}")]
-    ParseNice(&'static str),
+    ParseNice(String),
```
Comment (timvisee) [inline lib/common/common/src/cpu.rs:211]: Why is this necessary? Due to a dependency bump?

File: lib/segment/src/payload_storage/mmap_payload_storage.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            Ok(())
+        })
+    }
+
+    fn iter<F>(&self, mut callback: F) -> OperationResult<()>
+    where
+        F: FnMut(PointOffsetType, &Payload) -> OperationResult<bool>,
+    {
+        self.storage.read().iter(|point_id, payload| {
+            match callback(point_id, payload) {
```
Comment (timvisee) [inline lib/segment/src/payload_storage/mmap_payload_storage.rs:154]: How about `callback(...).map_err(|e| std::io::Error::new(...))` here?

File: Cargo.lock (hunk 1 above)
Comment (coszio) [reply Cargo.lock:7]: I did 😬

File: lib/common/common/src/cpu.rs (hunk 4 above)
Comment (coszio) [reply lib/common/common/src/cpu.rs:211]: I assume it was a dependency bump, yes. But let's see if we can avoid it by not updating the whole project

File: lib/segment/Cargo.toml [hunk 6]
```diff
@@ -88,6 +88,8 @@ zerocopy = { workspace = true }
 sysinfo = "0.32"
 charabia = { version = "0.9.1", default-features = false, features = ["greek", "hebrew", "thai"] }
 
+blob_store = { git = "https://github.com/qdrant/mmap-payload-storage" }
```
Comment (agourlay) [inline lib/segment/Cargo.toml:91]: we just should lock the revision like we do for the `wal` crate

File: lib/segment/src/payload_storage/mmap_payload_storage.rs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            Ok(Self { storage })
+        } else {
+            Err(OperationError::service_error(
+                "Failed to open mmap payload storage",
+            ))
+        }
+    }
+
+    fn new(path: PathBuf) -> Self {
+        let storage = BlobStore::new(path, StorageOptions::default()).unwrap();
```
Comment (agourlay) [inline lib/segment/src/payload_storage/mmap_payload_storage.rs:59]: this unwrap should go :)

File: lib/segment/src/payload_storage/mmap_payload_storage.rs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            // create folder if it does not exist
+            std::fs::create_dir_all(&path).map_err(|_| {
+                OperationError::service_error("Failed to create mmap payload storage directory")
+            })?;
+            Ok(Self::new(path))
+        }
+    }
+
+    fn open(path: PathBuf) -> OperationResult<Self> {
+        if let Some(storage) = BlobStore::open(path) {
```
Comment (agourlay) [inline lib/segment/src/payload_storage/mmap_payload_storage.rs:48]: we should have an error here explaining exactly why it can't opened

Comment (coszio) [reply lib/segment/src/payload_storage/mmap_payload_storage.rs]: https://github.com/qdrant/qdrant/pull/5398/commits/8f697c26c0e14aa7c417c58c1bcee5e260f11431

File: lib/segment/src/payload_storage/mmap_payload_storage.rs (hunk 7 above)
Comment (coszio) [reply lib/segment/src/payload_storage/mmap_payload_storage.rs]: https://github.com/qdrant/qdrant/pull/5398/commits/8f697c26c0e14aa7c417c58c1bcee5e260f11431

File: lib/segment/Cargo.toml (hunk 6 above)
Comment (coszio) [reply lib/segment/Cargo.toml]: https://github.com/qdrant/qdrant/pull/5398/commits/8f697c26c0e14aa7c417c58c1bcee5e260f11431

File: lib/segment/src/payload_storage/mmap_payload_storage.rs (hunk 5 above)
Comment (coszio) [reply lib/segment/src/payload_storage/mmap_payload_storage.rs:165]: Yeah, better

File: lib/segment/src/payload_storage/tests.rs [hunk 9]
```diff
@@ -0,0 +1,106 @@
+use serde_json::json;
+
+use super::mmap_payload_storage::MmapPayloadStorage;
+use super::on_disk_payload_storage::OnDiskPayloadStorage;
+use super::simple_payload_storage::SimplePayloadStorage;
+use super::PayloadStorage;
+use crate::common::rocksdb_wrapper::open_db;
+use crate::types::Payload;
+
+fn test_trait_impl<S: PayloadStorage>(mut storage: S) {
```
Comment (agourlay) [inline lib/segment/src/payload_storage/tests.rs:10]: WDYT about adding `open` to the `PayloadStorage` interface?

This way we could test persistence here by dropping the storage and  then reopening it to check that the payloads are still present.

**Comment by agourlay** [comment]
Looking very good, one last ask to improve testing then it is good to go for me :+1:

**Review by agourlay — APPROVED** [review]

**Review by timvisee — APPROVED** [review]
